Added ability to import multiple files.

n1snt opened the pull request:
## Purpose / Description
Implemented the feature to import multiple files at once.  #10641 

## Approach
What I have done to implement this feature:
- Enabled `EXTRA_ALLOW_MULTIPLE` in `openImportFilePicker` in `ImportFileSelectionFragment`. This enables the filepicker to select multiple files.
- Moved to using ArrayList instead of String in `handleFileImportInternal` in `ImportUtils` as the application will be handling more than one file.
- `handleContentProviderFile` now accepts an ArrayList of Uri instead of a string for path.
- `sendShowImportFileDialogMsg` now accepts ArrayList instead of a String.
- `ImportDialog` also uses StringArrayList instead of string & then it iterates over it and displays a list of files that will be imported (Screenshot below).
- `ImportAdd` takes the ArrayList and iterates over it to import files.
- `ImportReplace` also takes ArrayList as an argument instead of string.

## Screenshots:
<p align="center">
<img src="https://user-images.githubusercontent.com/46300244/163457916-9d9e4d5b-2177-49bf-bb12-75a14cc8d8d0.jpg" width="49%">  <img src="https://user-images.githubusercontent.com/46300244/163457921-97884ff0-cc8a-45bc-832d-e6d5b7bbab0e.jpg" width="49%">
</p>

## How Has This Been Tested?

I have tested my changes on:
- Physical device (Android 12.1)
- Virtual Device (Android 10)
- Ran all the tests

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented on your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison commented on .idea/codeStyles/Project.xml:
```diff
@@ -51,26 +51,8 @@
-          <package name="" alias="true" withSubpackages="true" />
-        </value>
-      </option>
+      <option name="CODE_STYLE_DEFAULTS" value="KOTLIN_OFFICIAL" />
```
This should be reverted

n1snt replied:
Done

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ImportDialog.kt:
```diff
@@ -36,28 +36,38 @@ class ImportDialog : AsyncDialogFragment() {
+        var fileNameCount = 0
+        dialogMessageList.forEach {
+            fileNameCount += 1
+            displayFileName += fileNameCount
+            displayFileName += ". "
+            displayFileName += filenameFromPath(convertToDisplayName(it))
+            displayFileName += "\n"
```
Concatenation means our translators can't change strings. Aim for string formatting when possible, and make the format string translatable)

n1snt replied:
I'll try to do this without concatenation.

n1snt replied:
Hey, I looked into this & came to the conclusion that it cannot be made to work without concatenation & I don't think it needs to be translatable because what this code does is display the list of selected files like this:
<img src="https://user-images.githubusercontent.com/46300244/163521355-17498823-e048-4c5f-985c-bd7009914b1c.jpg" width="49%">

mikehardy replied:
This is not correct. You are forgetting RTL languages (Hebrew, Arabic...)
We support those and do not want a regression for them
Each line needs to be a formatted screen, but you can concatenate multiple lines together

n1snt replied:
Yeah, I just didn't think about RTL.
So I'm still a bit confused what isn't RTL compliant here?
The new lines I'm concatenating?
Or the numbering I have done?
(Sorry I have never worked with RTL before. 😅)

n1snt replied:
How is this?
After:
![image](https://user-images.githubusercontent.com/46300244/163592106-1b4bd4bb-b016-43c1-8f78-bae9ea3ea87b.png)
![image](https://user-images.githubusercontent.com/46300244/163592173-f92357da-87c4-466a-9b2b-2ef3c2c3bcc8.png)

mikehardy replied:
I don't mind the number being there or not there, but that was the problem. You were concatenating a string with components and the number would have to go on the opposite side for RTL
Without the number, that can work, and doesn't require translation as long as each item is on a separate line. If there are 10, does it scroll?

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1091,35 +1091,44 @@ protected StudyOptionsFragment.DeckStudyData task(@NonNull Collection col, @NonN
+
+            ArrayList<AnkiPackageImporter> impList;
+            // Iterate over mPath
+            for(int i=0; i<= mPath.size(); i++){
+
+                imp = new AnkiPackageImporter(col, mPath.get(i));
+                imp.setProgressCallback(new TaskManager.ProgressCallback(collectionTask, res));
+                try {
+                    imp.run();
+                } catch (ImportExportException e) {
+                    Timber.w(e);
+                    return new Triple(null, true, e.getMessage());
+                }
+            };
```
https://github.com/ankidroid/Anki-Android/wiki/Code%20style
```suggestion
            for(int i = 0; i<= mPath.size(); i++) {

                imp = new AnkiPackageImporter(col, mPath.get(i));
                imp.setProgressCallback(new TaskManager.ProgressCallback(collectionTask, res));
                try {
                    imp.run();
                } catch (ImportExportException e) {
                    Timber.w(e);
                    return new Triple(null, true, e.getMessage());
                }
            };
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1091,35 +1091,44 @@ protected StudyOptionsFragment.DeckStudyData task(@NonNull Collection col, @NonN
-                return new Triple(null, true, e.getMessage());
-            }
+
+           AnkiPackageImporter imp = null; 
```
```suggestion
            AnkiPackageImporter imp = null; 
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1091,35 +1091,44 @@ protected StudyOptionsFragment.DeckStudyData task(@NonNull Collection col, @NonN
+                    imp.run();
+                } catch (ImportExportException e) {
+                    Timber.w(e);
+                    return new Triple(null, true, e.getMessage());
```
What does this mean? We return a failure even if there was a success - this probably wants to handle the case that some succeed and some fail

n1snt replied:
While iterating over mPath and importing the files if an exception gets raised it will return that error.
If no exceptions were raised during iteration then it will return the Triple with imp (variable) instead of the null which is passed when there is an error.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1129,133 +1138,138 @@ protected Computation<?> task(@NonNull Collection col, @NonNull ProgressSenderAn
-                BackupManager.removeDir(dir);
-            }
+            // Iterate over mPath list.
+            for(int i=0; i < mPath.size(); i++){
```
```suggestion
            for (int i = 0; i < mPath.size(); i++) {
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1129,133 +1138,138 @@ protected Computation<?> task(@NonNull Collection col, @NonNull ProgressSenderAn
-                if (zip.getEntry(colname) == null) {
-                    colname = CollectionHelper.COLLECTION_FILENAME;
+                // extract the deck from the zip file
+                String colPath = CollectionHelper.getCollectionPath(context);
```
I need to head out, this is a partial review up to here.

Overall, this looks great so far!

david-allison reviewed (changes requested):
I need to head out, this looks great! Partial review

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -1964,15 +1965,15 @@ public void loginToSyncServer() {
     // Callback to import a file -- replacing the existing collection
     @Override
-    public void importReplace(String importPath) {
+    public void importReplace(ArrayList<String> importPath) {
```
rename to importPaths

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -125,65 +126,80 @@ object ImportUtils {
+            val tempOutDirList: ArrayList<String> = ArrayList()
+
+            // Iterate over data & append tempOutDir to above arraylist.
+            data.forEach {
```
style (optional): I'd name `it` here

n1snt replied:
I should rename the default `it` to something else?

david-allison replied:
Yep

david-allison replied:
still pending. rename to `data`

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -303,15 +319,16 @@ object ImportUtils {
+
                 // Get the filename from the path
-                val f = File(path)
+                val f = File(pathList[0])
```
```suggestion
                val f = File(pathList.first())
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -303,15 +319,16 @@ object ImportUtils {
              * Send a Message to AnkiDroidApp so that the DialogMessageHandler shows the Import apkg dialog.
              * @param path path to apkg file which will be imported
```
This needs to be changed

david-allison commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -149,7 +149,7 @@
     <string name="import_message_add" comment="Adding a new deck to the collection (import)">Add</string>
     <string name="import_message_replace_confirm">This will delete your existing collection and replace it with the data of file %s</string>
-    <string name="import_message_add_confirm">Add “%s” to collection? This may take a long time</string>
+    <string name="import_message_add_confirm">Do you want to import the following files: \n%sThis may take a long time</string>
```
If you materially change a string, also change the key (only in this file) and change references to it. This forces a retranslation

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/CardBrowser_CardCollectionTest.kt:
```diff
@@ -42,7 +42,22 @@ class CardBrowser_CardCollectionTest {
     }
 
-    private class Positioned(override var position: Int) : PositionAware {
+    @KotlinCleanup("See if we can have the variable override")
+    class Positioned(override var position: Int) : PositionAware {
         val initialValue: Int = position
+
+        @JvmName("getPosition1")
+        fun getPosition(): Int {
+            return position
+        }
+
+        @JvmName("setPosition1")
+        fun setPosition(value: Int) {
+            position = value
+        }
+
```
This can now be removed if you rebase onto `main`

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/CardBrowser_CardCollectionTest.kt:
```diff
@@ -42,7 +42,22 @@ class CardBrowser_CardCollectionTest {
+            val initialValue: Int = position
+        }
     }
 }
```
This can be reverted

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/BackupManagerIntegrationTest.kt:
```diff
@@ -38,7 +38,8 @@ class BackupManagerIntegrationTest : RobolectricTest() {
         @Suppress("UNUSED_VARIABLE")
         val unusedCol = col
-        val path = createBackup()
+        val path: ArrayList<String> = ArrayList<String>()
+        path.add(createBackup())
```
maybe:

```kotlin
        val path = ArrayList<String>(listOf(createBcackup()))
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -1397,12 +1398,12 @@ private void showSyncLogMessage(@StringRes int messageResource, String syncMessa
     }
 
-    public void showImportDialog(int id, String message) {
+    public void showImportDialog(int id, ArrayList<String> messageList) {
```
I'd consider using `List<>` rather than `ArrayList<>` unless the list needs to be mutable

david-allison reviewed (changes requested):
I'd consider using `List<>` rather than `ArrayList<>` throughout.

After this is done, we can use a Kotlin `MutableList` in the tests (and later change this to `List<>` in the code), which will reduce the amount of test code.

Could you add a few `@NeedsTest()` attributes listing the tests which should be added:

* Test two files - success
* Test one failure, one success

Overall, this looks great!

n1snt commented:
Yes, it's scrollable @mikehardy 

![photo_2022-04-15_23-43-41](https://user-images.githubusercontent.com/46300244/163606190-26e56b6a-066b-422b-865f-06fa3317f187.jpg)

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -1558,7 +1560,7 @@ public void handleDbLocked() {
 
     public void restoreFromBackup(String path) {
-        importReplace(path);
+        importReplace(new ArrayList());
```
How does this work now the path isn't used?

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -125,65 +126,80 @@ object ImportUtils {
             return if (intentUriScheme == ContentResolver.SCHEME_CONTENT || intentUriScheme == ContentResolver.SCHEME_FILE || intentUriScheme == ContentResolver.SCHEME_ANDROID_RESOURCE) {
                 Timber.i("Attempting to read content from intent.")
-                handleContentProviderFile(context, intent, intent.data)
+                val intentDataList: ArrayList<Uri> = ArrayList()
+                intentDataList.add(intent.data!!)
+                handleContentProviderFile(context, intent, intentDataList)
```
might be able to reduce with `arrayListOf`

david-allison commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -149,7 +149,7 @@
     <string name="import_message_add" comment="Adding a new deck to the collection (import)">Add</string>
     <string name="import_message_replace_confirm">This will delete your existing collection and replace it with the data of file %s</string>
-    <string name="import_message_add_confirm">Add “%s” to collection? This may take a long time</string>
+    <string name="import_files_message_add_confirm">Do you want to import the following files: \n%sThis may take a long time</string>
```
This needs to be in a `plurals` XML element to ensure that translators can currently translate the number of files

https://developer.android.com/guide/topics/resources/string-resource#Plurals

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerImportTest.java:
```diff
@@ -37,7 +39,9 @@ public class DeckPickerImportTest extends RobolectricTest {
         DeckPickerImport deckPicker = super.startActivityNormallyOpenCollectionWithIntent(DeckPickerImport.class, new Intent());
 
-        deckPicker.showImportDialog(ImportDialog.DIALOG_IMPORT_ADD_CONFIRM, "");
+        ArrayList<String> messageList = new ArrayList<>();
+        messageList.add("");
+        deckPicker.showImportDialog(ImportDialog.DIALOG_IMPORT_ADD_CONFIRM, messageList);
```
Add to the top of the method: `@KotlinCleanup("arrayListOf")`

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerImportTest.java:
```diff
@@ -46,7 +50,9 @@ public void importAddShowsImportDialog() {
         DeckPickerImport deckPicker = super.startActivityNormallyOpenCollectionWithIntent(DeckPickerImport.class, new Intent());
 
-        deckPicker.showImportDialog(ImportDialog.DIALOG_IMPORT_REPLACE_CONFIRM, "");
+        ArrayList<String> messageList = new ArrayList<>();
+        messageList.add("");
+        deckPicker.showImportDialog(ImportDialog.DIALOG_IMPORT_REPLACE_CONFIRM, messageList);
```
Add to the top of the method: `@KotlinCleanup("arrayListOf")`

david-allison reviewed (commented):
Looks great! Few nitpicks, one plural change and it's great!

BrayanDSO commented:
One question: what happens if there is one or more `.colpkg` file on the selected files?

n1snt commented:
I need to look into it, I think it would need to add some checks while importing the files or limit the file picker by allowing only certain types of extensions.

mikehardy commented:
I'm not sure it is possible to say "allow multiple .apkg but only one .colpkg" as a configuration for the picker?
So I think it would have to be a post-processing step where before you do the import you verify there is only one .colpkg
Great catch @BrayanDSO

n1snt commented:
There are 2 options in the import dialog:
![image](https://user-images.githubusercontent.com/46300244/164012285-9b1e199d-7a32-4d78-8fc0-f649004b2ad6.png)
Both of the options allow selecting apkg/colpkg files.
Can't we restrict the apkg option to select only apkg files and the colpkg option to only colpkg files?

mikehardy commented:
Oh of course @n1snt - I did not realize we already differentiate between them before showing the picker. Restricting on file extension should be possible then yes

BrayanDSO commented:
Android distinguish files by MIME types, not extensions.

And as I said, Android MIME types can't tell the difference about a .colpkg and a .apkg. If you tap `Deck package (.apkg)`, you can still import a `.colpkg`. We can't even filter those type of files (#10413) since the MIME type may vary, so we use `*/*` as MIME type.

So after the files are selected, if there is a `.colpkg` file among them, no other files can be imported (both .apkg and .colpkg) since it replaces the whole collection. 

There must be an Anki method to distinguish `.colpkg` and `.apkg` after the files are selected, so I would go this way and tell the user if there is a `.colpkg` file in a multi-file import

n1snt commented:
If the user selects .apkg files & a .colpkg file then we should give users the option to select either the apkg files or the colpkg file.
I can try to make this work.

BrayanDSO commented:
As images, videos and any other file can be picked too, since the MIME is `*/*`, they should be flagged as invalid imports.

Maybe you could just add them on the summary you proposed [here](https://github.com/ankidroid/Anki-Android/issues/10641#issuecomment-1101659438) as something like `Invalid files: file1.ext, file2.ext`

n1snt commented:
@david-allison please check & let me know if there are any more changes to be made.
Now I'm working on adding checks after the user selects multiple files. I am implementing the following checks:
- If the user adds multiple colpkg files show a warning.
- If the user adds unsupported files with supported files. Show a dialog & allow the user to remove the unsupported file from the List & import the other files.
- If colpkg & apkg file(s) is/are selected at once then show a dialog & allow the user to select to either import the colpkg file or the apkg file(s)

Are there any other checks I should keep in mind? @BrayanDSO

BrayanDSO commented:
> Now I'm working on adding checks after the user selects multiple files. I am implementing the following checks

The plan looks great to me 👍

shunyadotzero commented:
Hi, I am very much in need of this feature of importing multiple files to form flashcard for each file separately in ankidroid.
I just want to know that can I use this feature somehow now

david-allison commented:
I like the plan.

Is there a possibility that we can get the feature in as-is, with additional functionality to handle selection of 'bad' files in follow-up PRs?

This is already fairly big and looks good overall.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -216,8 +217,8 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
 
-    private class ImportAddListener(deckPicker: DeckPicker?) : TaskListenerWithContext<DeckPicker, String, Triple<AnkiPackageImporter?, Boolean, String?>>(deckPicker) {
-        override fun actualOnPostExecute(context: DeckPicker, result: Triple<AnkiPackageImporter?, Boolean, String?>) {
+    private class ImportAddListener(deckPicker: DeckPicker?) : TaskListenerWithContext<DeckPicker, String, Triple<List<AnkiPackageImporter>?, Boolean, String?>>(deckPicker) {
```
suggest: moving away from a `Triple` to a Kotlin `class` for readability. There are dependencies between the parameters and this would improve things

Now we can make classes in a line of code, it doesn't seem worthwhile to have.

Please feel free to add a `@KotlinCleanup` instead and add an issue for me to take this one

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1616,8 +1631,8 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
                                 dialogMessage += """
-                                    
-                                    
+
```
revert spacing change

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1616,8 +1631,8 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
-                                    
-                                    
+
+
```
ditto

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1515,7 +1530,7 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
                     """
     $mCurrentMessage
-    
+
```
ditto

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1366,8 +1380,9 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
 
-    fun restoreFromBackup(path: String) {
-        importReplace(path)
+    fun restoreFromBackup(path: String?) {
+        val pathList: List<String> = listOfNotNull(path)
```
```suggestion
    fun restoreFromBackup(path: String) {
        val pathList: List<String> = listOf(path)
```

david-allison replied:
not handled

david-allison replied:
path was non-null, why was it made null?

n1snt replied:
I think I made it non-nullable accidentally. I have now made it the way it was before & now listOf() can be used here.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1681,7 +1696,7 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
                     val message = """
                         ${res.getString(R.string.sync_database_acknowledge)}
-                        
+
```
ditto

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -2456,7 +2473,7 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
                 """
      $dialogMessage
-     
+
```
spacing change

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ImportDialog.kt:
```diff
@@ -36,28 +36,33 @@ class ImportDialog : AsyncDialogFragment() {
         val builder = MaterialDialog.Builder(requireActivity())
         builder.cancelable(true)
-        val dialogMessage = requireArguments().getString("dialogMessage")!!
+        val dialogMessageList = requireArguments().getStringArrayList("dialogMessage")!!
```
this newline change doesn't seem necessary

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ImportDialog.kt:
```diff
@@ -36,28 +36,33 @@ class ImportDialog : AsyncDialogFragment() {
+        val dialogMessageList = requireArguments().getStringArrayList("dialogMessage")!!
+
+        // Iterate over dialog message list & create display filename
+        var displayFileName = ""
+        dialogMessageList.forEach {
+            displayFileName += filenameFromPath(convertToDisplayName(it)) + "\n"
+        }
```
```suggestion
        val displayFileName = dialogMessageList.joinToString("\n") { filenameFromPath(convertToDisplayName(it)) }
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:
```diff
@@ -783,146 +784,157 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
+                    imp.run()
+                } catch (e: ImportExportException) {
+                    Timber.w(e)
+                    return Triple(null, true, e.message)
```
Q: Do we want to return `impList` here: If we have 5 successes and 1 failure, we want to understand both

david-allison replied:
unaddressed

david-allison replied:
Can we add this s a separate issue then?

n1snt replied:
Yeah, I think it should have a separate issue of its own.

n1snt replied:
I have an idea to implement this without rewriting everything. 
I'll try to implement this by tomorrow.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:
```diff
@@ -783,146 +784,157 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
-            if (dir.exists()) {
-                BackupManager.removeDir(dir)
-            }
+            pathList.forEach {
```
style: I do prefer `for (path in pathList) {`. We lose the context of what `it` is in longer lambdas

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -125,65 +126,79 @@ object ImportUtils {
-                    return ImportResult.fromErrorString(AnkiDroidApp.getAppResources().getString(R.string.import_error_content_provider, AnkiDroidApp.getManualUrl() + "#importing"))
+            val tempOutDirList: ArrayList<String> = ArrayList()
+
+            dataList.forEach {
```
for (data in dataList)

david-allison commented on AnkiDroid/src/main/res/values/03-dialogs.xml:
```diff
@@ -98,7 +98,12 @@
+
+    <plurals name="import_complete_message">
+        <item quantity="one">Cards imported: %d</item>
+        <item quantity="other">Files imported: %d\nTotal cards imported: %d</item>
```
These both need to use positional string arguments: `%1$d` and `%2$d`

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -228,8 +229,22 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
+                val fileCount = imp?.size
+
+                context.showSimpleMessageDialog(
+                    fileCount?.let {
```
What happens if `fileCount` is null here? Are we passing in a null dialog?

david-allison replied:
What happens if `fileCount` is null here? Are we passing in a null dialog?

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/BackupManagerIntegrationTest.kt:
```diff
@@ -38,7 +38,7 @@ class BackupManagerIntegrationTest : RobolectricTest() {
         @Suppress("UNUSED_VARIABLE")
         val unusedCol = col
-        val path = createBackup()
+        val path: ArrayList<String> = ArrayList<String>(listOf(createBackup()))
```
`arrayListOf`

david-allison replied:
I think

```suggestion
        val path = arrayListOf(createBackup())
```

david-allison reviewed (changes requested):
This is all really good, mostly nitpicks.

Please, feel empowered to say 'no/not now' to any of the style suggestions. I'd like to see this in, and the code looks good.

One question on how we handle some successes and some failures

I wrote the code snippets without an IDE, so can you paste them in rather than assuming they compile

n1snt commented:
> 

Hey, I have almost implemented the error handling mentioned by @BrayanDSO . I will make the changes you mentioned as soon as I get back home.

n1snt commented:
Hey, @david-allison please check this out & let me know of any changes I need to make.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -228,8 +230,12 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
                 Timber.i("Import: Add succeeded")
-                val imp = result.first!!
-                context.showSimpleMessageDialog(TextUtils.join("\n", imp.log))
+                val imp = result.first
+
+                val fileCount = imp?.size
+                var totalCardCount = 0
+                imp?.forEach { totalCardCount += it.cardCount }
+                context.showSimpleMessageDialog(fileCount?.let { context.resources.getQuantityString(R.plurals.import_complete_message, it, fileCount, totalCardCount) })
                 context.updateDeckList()
```
```suggestion
                result.first?.let { importers ->
                    val fileCount = importers.size
                    val totalCardCount = importers.sumOf { it.cardCount }
                    context.showSimpleMessageDialog(
                            context.resources.getQuantityString(R.plurals.import_complete_message, fileCount, fileCount, totalCardCount)
                    )
                }
               context.updateDeckList()
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -903,7 +909,7 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
                     "message",
                     """
      ${res.getString(R.string.full_sync_confirmation_upgrade)}
-     
```
revert

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1616,8 +1622,6 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
                             dialogMessage = res.getString(R.string.sync_connection_error)
                             if (result.size >= 0 && result[0] is Exception) {
                                 dialogMessage += """
```
line changes

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -2456,7 +2461,6 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
             return if (!TextUtils.isEmpty(dialogMessage) && !TextUtils.isEmpty(syncMessage)) {
                 """
      $dialogMessage
```
line changes

david-allison reviewed (changes requested):
quite a few of my previous comments are unaddressed

Arthur-Milchior commented:
> Hi, I am very much in need of this feature of importing multiple files to form flashcard for each file separately in ankidroid. I just want to know that can I use this feature somehow now
@shunyadotzero if you manually install alphas, hopefully in a few weeks at most. Maybe days even. 
If you don't do manual installation, not before months as we can't yet update ankidroid and will need time to test once we can release alpha and beta

Arthur-Milchior commented:
I love the idea @n1snt. 
I know you were already asked a lot of changes in review, and the one I'm asking for is not that important comparatively (but it would help review the code and may be an interesting skills to learn. If you need help, I'm fine doing a pair programming session to attack it)
Ideally, I'd really appreciate if you can avoid having 60 commits. Use `git rebase --interactive HEAD^60` to fix them in a bunch of atomic commits so that it gets easier to review.

One way to create the commits, for example, would be to first modify `handleContentProviderFile` to take a list of URL, and the called would be in charge of wrapping its only element in a list.
Then a second commit would change handleFileImportInternal to ensure it deals with the list of urls and pass it to handlleContentProvideFile

One way it would help, for example, is that it would force to decide at which step of the process you need to add mCardCount to Importer. Right now, I have no idea why it was necessary to add a new property and remove the logging.

Another idea that would be very simple to review with atomic commits and a nightmare otherwise would be to split ImportReplace's  task into multiple pieces.
For example, we can extract some part of the code and put it into separate function, that would be called in `task`. Then a next change can ensure that the constructor takes a list instead of a single path, and iterate in a small commit which "just" call the function we created in the previous commits onto all of the elements of the list. This will avoid to have a huge part of code in an extra layer of imbrication, and a commit that appears to touch a lot of code while actually it makes a relatively simple changes.

Also, it would ensure that a lot of revert request are handled perfectly, since the commit in which extra lines and other errors were added would just be deleted

n1snt commented:
Hey, @david-allison I have just made the changes you suggested. I have also given reasons why I haven't made some of the changes. Check this out & let me know if there are any more changes to be made.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -2456,7 +2460,6 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
             return if (!TextUtils.isEmpty(dialogMessage) && !TextUtils.isEmpty(syncMessage)) {
                 """
      $dialogMessage
-     
```
line change

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -125,65 +126,103 @@ object ImportUtils {
+                    }
                 }
             }
-            return if (!isValidPackageName(filename)) {
-                if (isAnkiDatabase(filename)) {
-                    // .anki2 files aren't supported by Anki Desktop, we should eventually support them, because we can
-                    // but for now, show a "nice" error.
-                    ImportResult.fromErrorString(context.resources.getString(R.string.import_error_load_imported_database))
-                } else {
-                    // Don't import if file doesn't have an Anki package extension
-                    ImportResult.fromErrorString(context.resources.getString(R.string.import_error_not_apkg_extension, filename))
-                }
-            } else {
-                // Copy to temporary file
-                filename = ensureValidLength(filename)
-                val tempOutDir = Uri.fromFile(File(context.cacheDir, filename)).encodedPath!!
-                val errorMessage = if (copyFileToCache(context, data, tempOutDir)) null else context.getString(R.string.import_error_copy_file_to_cache)
-                // Show import dialog
-                if (errorMessage != null) {
-                    AnkiDroidApp.sendExceptionReport(RuntimeException("Error importing apkg file"), "IntentHandler.java", "apkg import failed")
-                    return ImportResult.fromErrorString(errorMessage)
+
+            if (apkgCount > 0 && colpkgCount > 0) {
+                Timber.i("Both apkg & colpkg selected.")
+                return ImportResult.fromErrorString(context.resources.getString(R.string.import_error_colpkg_apkg))
+            } else if (colpkgCount > 1) {
+                Timber.i("Multiple colpkg files selected.")
+                return ImportResult.fromErrorString(context.resources.getString(R.string.import_error_multiple_colpkg))
+            }
```
extract to a method as none of these variables are used again

```kotlin
validateImportTypes(dataList)?.let { errorMessage ->
    return ImportResult.fromErrorString(errorMessage)
}
```

david-allison commented on AnkiDroid/src/main/res/values/03-dialogs.xml:
```diff
@@ -98,7 +98,12 @@
-    <string name="import_complete_count">Cards imported: %d</string>
+
+    <plurals name="import_complete_message">
+        <item quantity="one">Cards imported: %d</item>
```
I presume: `%2$d`, %d seems to be the first param (file count)

n1snt replied:
Done

david-allison reviewed (changes requested):
Few minor changes, then good to go

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1681,7 +1684,6 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
                     // Note: Do not log this data. May contain user email.
                     val message = """
```
still has line changes:

1664 and 1619-1620 on the original

david-allison commented:
Let's get reviewer 2 involved, spacing changes aren't enough to hold this up

n1snt commented:
Now if an error occurs while importing multiple files it will continue trying to import all the files then at the import stats page it will show the files with the errors:
![image](https://user-images.githubusercontent.com/46300244/165846309-701ea73c-a19c-45d2-b9bf-9ef546b0521d.png)

n1snt commented:
@david-allison please check this out and let me know if there are any changes that need to be made.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -1358,7 +1383,8 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
 
     fun restoreFromBackup(path: String) {
-        importReplace(path)
+        val pathList: List<String> = listOf(path)
+        importReplace(pathList)
```
```suggestion
        importReplace(listOf(path))
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:
```diff
@@ -783,146 +784,162 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
+
+                imp.setProgressCallback(TaskManager.ProgressCallback(collectionTask, res))
+                try {
+                    if (impList == null) {
+                        impList = arrayListOf()
+                    }
```
Why not: initalize `impList` as `arrayListOf<AnkiPackageImporter>()` and make it non-null

david-allison replied:
We should make this explicit in the code via different returns

n1snt replied:
I'm thinking of a way to do this but that may require some significant changes.
It used to work the same way before just without the ArrayList stuff.
I think a different issue should be raised for this.

n1snt replied:
That would require moving to class from Triple as you suggested.

david-allison replied:
* issue: errFlag is never set

```kotlin
            val impList = arrayListOf<AnkiPackageImporter>()
            val errFlag = false
            var errList: String? = null

            for (path in pathList) {
                val imp = AnkiPackageImporter(col, path)

                imp.setProgressCallback(TaskManager.ProgressCallback(collectionTask, res))
                try {
                    impList.add(imp)
                    imp.run()
                } catch (e: ImportExportException) {
                    Timber.w(e)
                    errList += File(path).name + "\n" + e.message + "\n"
                }
            }

            return Triple(if (impList.isEmpty()) null else impList, errFlag, errList)
```

david-allison replied:
But, in general, I think there's an issue with the code structure here.

n1snt replied:
Wait I'll try the above changes.
errFlag was supposed to be set after `Timber.w(e)`
Yeah, I was adding imp to impList after `imp.run()` is executed. I think I accidentally moved it above.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -207,6 +223,39 @@ object ImportUtils {
+                errorMsg = context.resources.getString(R.string.import_error_colpkg_apkg)
+            } else if (colpkgCount > 1) {
+                Timber.i("Multiple colpkg files selected.")
+                errorMsg = context.resources.getString(R.string.import_error_multiple_colpkg)
```
```suggestion
                return context.getString(R.string.import_error_multiple_colpkg)
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -207,6 +223,39 @@ object ImportUtils {
+            var errorMsg: String? = null
+            if (apkgCount > 0 && colpkgCount > 0) {
+                Timber.i("Both apkg & colpkg selected.")
+                errorMsg = context.resources.getString(R.string.import_error_colpkg_apkg)
```
```suggestion
                return context.getString(R.string.import_error_colpkg_apkg)
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -207,6 +223,39 @@ object ImportUtils {
+                }
+            }
+
+            var errorMsg: String? = null
```
```suggestion
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -207,6 +223,39 @@ object ImportUtils {
+                errorMsg = context.resources.getString(R.string.import_error_multiple_colpkg)
+            }
+
+            return errorMsg
```
```suggestion
            return null
```

david-allison reviewed (approved):
Nothing blocking

david-allison reviewed (changes requested):
Further inspection on the unresolved comment

david-allison commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:
```diff
@@ -798,18 +798,16 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
+                    impList.add(imp)
                 } catch (e: ImportExportException) {
                     Timber.w(e)
+                    if (!errFlag) { errFlag = true }
```
```suggestion
                    errFlag = true
```

david-allison commented on AnkiDroid/src/androidTest/java/com/ichi2/anki/utils/DiffEngineTest.kt:
```diff
@@ -13,31 +13,25 @@
  * You should have received a copy of the GNU General Public License along with         *
  * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
  ****************************************************************************************/
+package com.ichi2.anki.utils
```
This shouldn't be in this PR

david-allison reviewed (changes requested):
```
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerImportTest.java:1: Error: All files in AnkiDroid must contain a GPLv3-compatible copyright header [MissingCopyrightHeader from com.ichi2.anki:lint-rules]
```

n1snt commented:
I have created another fork named multiple-file-imports, there I have cleaned up the git history as suggested by @Arthur-Milchior 
I have also implemented this from scratch to avoid any spacing issues.
Is there any way I can replace the main branch in this pull request with the multiple-file-imports branch? or do I need to create a new pull request altogether?

n1snt commented:
Hey, @david-allison please check this out & let me know if there are any changes to be made.

n1snt commented:
Hey @david-allison can you please check this out & let me know if there are any changes to be made?

david-allison commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -148,8 +148,12 @@
     <string name="import_message_add" comment="Adding a new deck to the collection (import)">Add</string>
     <string name="import_message_replace_confirm">This will delete your existing collection and replace it with the data of file %s</string>
-    <string name="import_message_add_confirm">Add “%s” to collection? This may take a long time</string>
+    <plurals name="import_files_message_add_confirm">
```
change the key to force a re-translation

n1snt replied:
It's already a different key.
Which key needs to be changed here?

david-allison replied:
🤦‍♂️  So it is! Looks good to me!

david-allison reviewed (changes requested):
~~Looks good! Really minor change, so marking it as waiting for approval~~ + needs second approval

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:
```diff
@@ -783,146 +783,160 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
+                } catch (e: ImportExportException) {
+                    Timber.w(e)
+                    errFlag = true
+                    errList += File(path).name + "\n" + e.message + "\n"
```
@KotlinCleanup("Use StringBuilder to concatenate the strings")

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DialogHandler.kt:
```diff
@@ -49,10 +49,10 @@ class DialogHandler(activity: AnkiActivity) : Handler(getDefaultLooper()) {
             // Handle import of collection package APKG
-            (mActivity.get() as DeckPicker?)!!.showImportDialog(ImportDialog.DIALOG_IMPORT_REPLACE_CONFIRM, msgData.getString("importPath"))
+            (mActivity.get() as DeckPicker?)!!.showImportDialog(ImportDialog.DIALOG_IMPORT_REPLACE_CONFIRM, msgData.getStringArrayList("importPath") as java.util.ArrayList<String>)
         } else if (msg.what == MSG_SHOW_COLLECTION_IMPORT_ADD_DIALOG) {
             // Handle import of deck package APKG
-            (mActivity.get() as DeckPicker?)!!.showImportDialog(ImportDialog.DIALOG_IMPORT_ADD_CONFIRM, msgData.getString("importPath"))
+            (mActivity.get() as DeckPicker?)!!.showImportDialog(ImportDialog.DIALOG_IMPORT_ADD_CONFIRM, msgData.getStringArrayList("importPath") as java.util.ArrayList<String>)
```
Please remove `import java.util.*` from the imports and `java.util.` from these lines, leaving only `ArrayList<String>`, so it uses Kotlin's ArrayList instead of Java's

n1snt replied:
Done

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -214,20 +215,45 @@ open class DeckPicker : NavigationDrawerActivity(), StudyOptionsListener, SyncEr
+                if (result.third != null) {
+                    errorMsg += result.third
+                }
+                if (errorMsg.isNotBlank() or errorMsg.isNotEmpty()) {
```
Correct me if I'm wrong, but I think that if the error message isn't blank, logically it isn't empty as well. so only one of these checks is necessary

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:
```diff
@@ -783,146 +783,160 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
-            zip = try {
-                ZipFile(File(path))
-            } catch (e: IOException) {
-                Timber.e(e, "doInBackgroundImportReplace - Error while unzipping")
-                CrashReportService.sendExceptionReport(e, "doInBackgroundImportReplace0")
-                return Computation.ERR
-            }
-            try {
-                // v2 scheduler?
-                if (zip.getEntry(colname) == null) {
-                    colname = CollectionHelper.COLLECTION_FILENAME
+                // from anki2.py
+                var colname = "collection.anki21"
```
These lines and probably some others below need to run only once on the method, so they can be extracted from the for loop

n1snt replied:
I have moved some code which only needed to run once outside the loop.

BrayanDSO commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -148,8 +148,12 @@
+        <item quantity="other">Do you want to import the following files:\n%s\nThis may take a long time</item>
+    </plurals>
     <string name="import_log_no_apkg">This isn’t a valid Anki package file</string>
+    <string name="import_title_error">Error!</string>
```
Hmmmm. I particularly don't like the exclamation mark on the title, but I'd like to hear others' opinions on the matter

n1snt replied:
Removed the exclamation mark.

BrayanDSO commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -164,6 +168,9 @@
     <string name="import_error_unhandled_scheme">Unhandled import from: “%s”</string>
+    <string name="import_error_multiple_colpkg">Multiple colpkg files were selected.</string>
+    <string name="import_error_colpkg_apkg">Both apkg &amp; colpkg files were selected at once.</string>
+    <string name="import_stats_error">\n\nSome errors occurred while importing the following files:\n%s</string>>
```
I believe the `\n\n` on the beginning would be better being added on the code instead of here. This would allow more flexibility to change the formatting if necessary

n1snt replied:
Moved it.

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt:
```diff
@@ -207,6 +221,38 @@ object ImportUtils {
+                }
+            }
+
+            if (apkgCount > 0 && colpkgCount > 0) {
+                Timber.i("Both apkg & colpkg selected.")
+                return context.resources.getString(R.string.import_error_colpkg_apkg)
+            } else if (colpkgCount > 1) {
+                Timber.i("Multiple colpkg files selected.")
+                return context.resources.getString(R.string.import_error_multiple_colpkg)
+            }
+
+            return null
```
This doesn't seem to handle `otherCount`. Is this intended?

n1snt replied:
Remove count was just one of the things I was testing.
I concluded it was not needed but I kept it.
I have now removed it.

BrayanDSO reviewed (commented):
Overall, LGTM. Left some comments on some points.

I noticed `EXTRA_ALLOW_MULTIPLE` is added to the common method of `.apkg` and `.colpkg` dialogs. This is intended to allow the whole functionality of importing multiple files.

Could you create an issue to make the `.colpkg` option not have this intent extra? So it would be more natural to be able to choose only one file there. 

Restore backup dialog's `Choose another` option uses the same import dialog code, so it should help on it too.

BrayanDSO reviewed (approved):
LGTM. Thanks for the great job!

Please remember of creating the issue to allow only single imports on the `.colpkg` option on the import dialog, after this is merged.

n1snt commented:
Sure I'll do it as soon as it gets merged. 👍

mikehardy reviewed (approved):
This is a heck of a feature, very nice thank you! Please do remember to log the related issue as mentioned in prior comments :-)
